File: modules/greedyalgorithms/offline_caching.py
"""
Offline Caching: Belady's MIN vs. LRU  (CSC310 Module 12 - Greedy Algorithms)
============================================================================

The OFFLINE caching problem knows the whole request sequence in advance.  The
greedy rule "on a miss with a full cache, evict the item whose next use is
farthest in the future" (Belady's MIN) is provably optimal -- it causes the
fewest possible misses.  An ONLINE policy like LRU cannot see the future, so it
generally suffers more misses.  This program runs both on the same input and
prints a miss-by-miss trace so you can see why MIN wins.
"""
from typing import List


def belady(requests: List[int], k: int):
    """Optimal offline policy: evict the page whose next use is farthest away."""
    cache, misses, trace = [], 0, []
    for t, page in enumerate(requests):
        if page in cache:
            trace.append((page, "hit", list(cache)))
            continue
        misses += 1
        if len(cache) < k:
            cache.append(page)
            trace.append((page, "miss/load", list(cache)))
        else:
            # choose the cached page used farthest in the future (or never)
            def next_use(p):
                for j in range(t + 1, len(requests)):
                    if requests[j] == p:
                        return j
                return float("inf")
            victim = max(cache, key=next_use)
            cache[cache.index(victim)] = page
            trace.append((page, f"miss/evict {victim}", list(cache)))
    return misses, trace


def lru(requests: List[int], k: int):
    """Online policy: evict the least-recently-used page."""
    cache, misses, trace = [], 0, []   # cache ordered oldest-use -> newest-use
    for page in requests:
        if page in cache:
            cache.remove(page)
            cache.append(page)
            trace.append((page, "hit", list(cache)))
            continue
        misses += 1
        if len(cache) < k:
            cache.append(page)
            trace.append((page, "miss/load", list(cache)))
        else:
            victim = cache.pop(0)
            cache.append(page)
            trace.append((page, f"miss/evict {victim}", list(cache)))
    return misses, trace


def show(name, misses, trace):
    print(f"=== {name}: {misses} misses ===")
    for page, action, cache in trace:
        print(f"  request {page:<2}  {action:<14}  cache={cache}")
    print()


def main():
    requests = [1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4, 5]
    k = 3
    print(f"Request sequence: {requests}")
    print(f"Cache size k = {k}\n")

    bm, bt = belady(requests, k)
    lm, lt = lru(requests, k)
    show("Belady's MIN (optimal, offline)", bm, bt)
    show("LRU (online)", lm, lt)

    print(f"Belady misses = {bm}, LRU misses = {lm}")
    print("Belady is optimal, so it never has more misses than LRU.")


if __name__ == "__main__":
    main()

File: modules/greedyalgorithms/offline_caching.cpp
// Offline Caching: Belady's MIN vs. LRU  (CSC310 Module 12 - Greedy Algorithms)
// ============================================================================
//
// Belady's MIN (offline, optimal): on a miss with a full cache, evict the page
// whose next use is farthest in the future.  LRU (online) evicts the
// least-recently-used page.  MIN is provably optimal, so it never has more
// misses than LRU.
//
// Build:  g++ -std=c++17 -O2 offline_caching.cpp -o offline_caching && ./offline_caching

#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using std::vector;

static void printCache(const vector<int>& c) {
    std::cout << "cache=[";
    for (size_t i = 0; i < c.size(); ++i)
        std::cout << c[i] << (i + 1 < c.size() ? ", " : "");
    std::cout << "]\n";
}

// Belady's MIN: evict the page whose next use is farthest in the future.
int belady(const vector<int>& req, int k) {
    vector<int> cache;
    int misses = 0;
    for (size_t t = 0; t < req.size(); ++t) {
        int page = req[t];
        if (std::find(cache.begin(), cache.end(), page) != cache.end()) {
            std::cout << "  request " << page << "  hit            ";
            printCache(cache);
            continue;
        }
        ++misses;
        if ((int)cache.size() < k) {
            cache.push_back(page);
            std::cout << "  request " << page << "  miss/load      ";
            printCache(cache);
        } else {
            int victimIdx = 0, farthest = -1;
            for (size_t c = 0; c < cache.size(); ++c) {
                int nextUse = std::numeric_limits<int>::max();
                for (size_t j = t + 1; j < req.size(); ++j)
                    if (req[j] == cache[c]) { nextUse = (int)j; break; }
                if (nextUse > farthest) { farthest = nextUse; victimIdx = (int)c; }
            }
            std::cout << "  request " << page << "  miss/evict " << cache[victimIdx] << "  ";
            cache[victimIdx] = page;
            printCache(cache);
        }
    }
    return misses;
}

// LRU: cache ordered oldest-use -> newest-use; evict the front on a full miss.
int lru(const vector<int>& req, int k) {
    vector<int> cache;
    int misses = 0;
    for (int page : req) {
        auto it = std::find(cache.begin(), cache.end(), page);
        if (it != cache.end()) {
            cache.erase(it);
            cache.push_back(page);
            std::cout << "  request " << page << "  hit            ";
            printCache(cache);
            continue;
        }
        ++misses;
        if ((int)cache.size() == k) {
            std::cout << "  request " << page << "  miss/evict " << cache.front() << "  ";
            cache.erase(cache.begin());
            cache.push_back(page);
        } else {
            cache.push_back(page);
            std::cout << "  request " << page << "  miss/load      ";
        }
        printCache(cache);
    }
    return misses;
}

int main() {
    vector<int> req = {1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4, 5};
    int k = 3;

    std::cout << "Request sequence: ";
    for (int p : req) std::cout << p << " ";
    std::cout << "\nCache size k = " << k << "\n\n";

    std::cout << "=== Belady's MIN (optimal, offline) ===\n";
    int bm = belady(req, k);
    std::cout << "\n=== LRU (online) ===\n";
    int lm = lru(req, k);

    std::cout << "\nBelady misses = " << bm << ", LRU misses = " << lm << "\n";
    std::cout << "Belady is optimal, so it never has more misses than LRU.\n";
    return 0;
}
